contracts/vaults/TreasuryVault/RibbonAutocallVault.sol

```diff
@@ -0,0 +1,444 @@
+     * @param _obsFreq is the observation period frequency
+     */
+    function setPeriodAndObservationFrequency(uint256 _period, uint256 _obsFreq)
+        external
+        onlyOwner
+    {
+        require(_period > 0, "!_period");
+        require(_obsFreq > 0 && (period * 1 days) % _obsFreq == 0, "!_obsFreq");
+
+        emit ObservationPeriodFreqSet(obsFreq, _obsFreq);
+
+        emit PeriodSet(period, _period);
+
+        nextObsFreq = _obsFreq;
+        nextPeriod = _period;
+    }
+
+    /**
+     * @dev Overrides RibbonTreasuryVault commitAndClose()
+     */
+    function commitAndClose() external override nonReentrant {
+        address currentOption = optionState.currentOption;
+
+        if (currentOption == address(0)) {
+            // Commit and close vanilla put
+            super._commitAndClose();
+            // Commit and close enhanced put
+            _commitAndCloseEnhancedPut(0, 0);
+            return;
+        }
+
+        IOtoken currentOToken = IOtoken(currentOption);
+        uint256 expiry = currentOToken.expiryTimestamp();
+
+        (uint256 autocallTS, uint256 nCBBreaches, uint256 lastCBBreach) =
+            _autocallState(expiry);
+
+        // If before expiry, attempt to autocall
+        if (block.timestamp < expiry) {
+            // Require autocall barrier hit at least once
+            require(autocallTS < block.timestamp, "!autocall");
+            // Burn the unexpired oTokens
+            _burnRemainingOTokens();
+            // Require vault possessed all oTokens sold to counterparties
+            require(vaultState.lockedAmount == 0, "!withdrawnCollateral");
+        }
+
+        // Commit and close vanilla put
+        super._commitAndClose();
+
+        // Commit and close enhanced put
+        _commitAndCloseEnhancedPut(expiry, currentOToken.strikePrice());
+
+        (, , uint256 returnAmt) = _couponsEarned(nCBBreaches, lastCBBreach);
+
+        // Transfer unearned coupons back to autocall seller
+        transferAsset(autocallSeller, returnAmt);
+
+        // Set coupon state
+        CouponState memory _couponState = couponState;
+        couponState.couponType = _couponState.nCouponType;
+        couponState.AB = _couponState.nextAB;
+        couponState.CB = _couponState.nextCB;
+
+        // Set observation period frequency
+        obsFreq = nextObsFreq;
+        period = nextPeriod;
+        nTotalObs = period / obsFreq;
+    }
+
+    /**
+     * @dev Settles the enhanced put
+     * @param _expiry is the expiry of the current option
+     * @param _strikePrice is the strike of the current option
+     */
+    function _commitAndCloseEnhancedPut(uint256 _expiry, uint256 _strikePrice)
+        internal
+    {
+        uint256 expiryPrice =
+            ORACLE.getExpiryPrice(vaultParams.underlying, _expiry);
+
+        PutOption memory _putOption = putOption;
+
+        // If put ITM, transfer to autocall seller
+        if (_putOption.payoff > 0 && expiryPrice <= _strikePrice) {
```

Comment:
If [this](https://github.com/ribbon-finance/ribbon-v2/pull/311/files#r1183053697) applies, then one idea could be to add in the if clause that `expiryPrice > 0`

Reply:
doesn't apply because otherwise commitAndClose would not settle since the expiry price for the vanilla put has not been set.